backend/app/matrix/background_validator.py

```diff
@@ -0,0 +1,242 @@
+
+        if self.github_username is not None:
+            gh = gidgethub.aiohttp.GitHubAPI(
+                app_state.http_client.session,
+                requester=self.github_username,
+                oauth_token=user_data.content.github.access_token,
+            )
+
+            self.github_api = GithubAPI(
+                gh=gh,
+                username=self.github_username,
+                default_role=app_state.settings.github.organisation_membership,
+            )
+
+
+class BackgroundValidater:
+    def __init__(self):
+        self.client = app_state.bot_client
+        self.registered_users: Dict[str, RegisteredUser] = dict()
+
+        # List of rooms registered under the application and has "room membership" conditions.
+        self.rooms_with_conditions: List[str] = []
+
+        # Set of rooms where the bot has invite and kick permissions.
+        self.rooms_with_permissions: Set[str] = {}
+        self.background_validator_task = None
+
+    async def create_background_task(self):
+        print("Background validator has started")
+        self.background_validator_task = asyncio.create_task(self.start_task())
+
+    async def cancel_background_task(self):
+        self.background_validator_task.cancel()
+        print("Background validator has stopped")
+
+    async def start_task(self):
+        while True:
+            await self.fetch_rooms_and_users()
+
+            for room_id in self.rooms_with_conditions:
+                # When the bot is either not part of the room or dosen't have the required permissions.
+                if room_id not in self.rooms_with_permissions or room_id not in self.client.rooms:
+                    continue
+
+                ignore_members = await self.get_users_not_removed_by_bot(room_id)
+                room_specific_data = await self.client.get_account_data("rooms", room_id=room_id)
+
+                for user_id in self.registered_users.keys():
+                    if user_id in ignore_members:
+                        continue
+
+                    current_state = await self.get_current_room_membership(user_id, room_id)
+                    if current_state == "ignore":
+                        continue
+                    print(user_id, current_state)
+
+                    is_permitted_through_github = await self.check_github_conditions(
+                        user_id, room_specific_data
+                    )
+                    is_permitted = is_permitted_through_github
+
+                    next_state = await self.decide_next_state_of_user(
+                        current_state, is_permitted, room_specific_data.content.disable_room_kick
+                    )
+
+                    if current_state == next_state:
+                        continue
+
+                    await self.send_next_state_event(room_id, user_id, next_state)
+
+                await asyncio.sleep(1)
+            await asyncio.sleep(1)
```

Comment:
Does this mean we check all the auth of all users every second?

Reply:
Yes, auth of all users for a particular room with a one-second delay between every room.

Reply:
I was trying to play around with the idea of adding delays, so we can increase the delay if it suits us.